Support glob syntax in workspace members

hjr3 opened the pull request:
Fixes #3911

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

matklad commented on src/cargo/core/workspace.rs:
```diff
@@ -316,9 +318,16 @@ impl<'cfg> Workspace<'cfg> {
+            }
+
+            for path in expanded_list {
+                let manifest_path = path.join("Cargo.toml");
```
Will this error if there's a path **without** `Cargo.toml` matched by the glob? I would say that it's OK to bail in this case, but that just skipping that path would be probably more useful: I often have various assorted folders alongside the workspace packages. 

In any case, I think a test case for this would be useful :)

hjr3 replied:
Good call out. I will get on it.

alexcrichton commented:
Thanks for the PR @hjr3! Looks good to me but I'd share the same concerns as @matklad, mayb ejust add a test to see what the error looks like?

alexcrichton commented:
Also mind updating the workspace documentation to mention that globs may be used?

hjr3 commented:
@alexcrichton sure, i can update the docs too!

nipunn1313 commented:
Thanks @hjr3! I'm looking forward to this one (we currently have a big setup with a script to autogenerate the workspace members).

hjr3 commented:
Apologies for taking so long to update this PR. I have updated the docs. Is it ok to reuse the `globs` URL like I did? It renders correctly with my testing.

I also added the test to shows what happens with a workspace path does not have a `Cargo.toml` file. As expected, it throws an error. I would expect that someone would add that to the `exclude` array though. Is that good enough? If it is, does the error message need to be improved? Should I mention that in the docs?

alexcrichton commented:
Looks good to me, but the tests seem to be failing? The new `invalid_members` test also seems to not have an error message :(

hjr3 commented:
@alexcrichton ok i fixed `glob_syntax_invalid_members` to have the proper expectations. The `invalid_members` test was failing because that member path does not actually exist and the glob will ignore it. If expanded the member path returns an empty list, I default to the original (non-expanded) value. That could be weird in some cases though if someone specifies a pattern that does not match anything. It might be better to return an error if the pattern does not match anything?

alexcrichton commented on src/cargo/core/workspace.rs:
```diff
@@ -527,6 +544,18 @@ impl<'cfg> Workspace<'cfg> {
 }
 
+fn expand_member_path(path: &Path) -> CargoResult<Vec<PathBuf>> {
+    let path = path.to_str().unwrap();
```
Could the `None` case be handled here instead of `unwrap`? (returning an empty `Vec`)

hjr3 replied:
fixed

alexcrichton commented on src/cargo/core/workspace.rs:
```diff
@@ -527,6 +544,18 @@ impl<'cfg> Workspace<'cfg> {
 
+fn expand_member_path(path: &Path) -> CargoResult<Vec<PathBuf>> {
+    let path = path.to_str().unwrap();
+    let res = glob(path).map_err(|e| {
```
Could this use `.chain_error` instead of `.map_err`?

hjr3 replied:
I may be missing something here, but I don't see how `.chain_error` can be used with `Result<glob::Paths, glob::PatternError>`.

alexcrichton commented on src/cargo/core/workspace.rs:
```diff
@@ -527,6 +544,18 @@ impl<'cfg> Workspace<'cfg> {
+        human(format!("could not parse pattern `{}`: {}", &path, e))
+    })?;
+    res.map(|p| {
+        p.or_else(|e| {
```
As above, could this use `chain_error`?

alexcrichton commented:
Nah this seems like a reasonable interpretation to me I think, thanks!

alexcrichton commented:
@hjr3 mind updating to using `chain_error` instead of `map_err` in a few locations as well?

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit f00c223 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton f00c223d3a5e35559934549b9a10e83d02f18e42 -->

hjr3 commented:
@alexcrichton thanks, i was just about to push this same patch. it took me some time to figure out how to get the `chain_error` working for glob.

bors commented:
:hourglass: Testing commit f00c223 with merge 828a9c5...

alexcrichton commented:
Oh sorry about that! I was just going through old Cargo PRs and figured this could use a bit of a boost  :)

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.1868), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/233021719?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 828a9c56ad0a3ff83ba25d7d9b66d3dcfa23e571 to master...

Astlaan commented:
How to use?

```
[workspace]

members = [
    "./*"
]
```

gives an error

hi-rustin commented:
@Astlaan What is the error? Could you please open a new issue instead of leaving a comment here, as this PR was closed 7 years ago.

If it is a usage issue, https://users.rust-lang.org/ may be a better place to ask this type of question.

weihanglo commented:
@Astlaan you can use [`workspace.exclude`](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields) to exclude folder not intended to be a member.

See https://github.com/rust-lang/cargo/issues/11405 for tracking the workspace globbing issue
